### PR TITLE
Add ApplicationInsights telemetry to track usage of Glimpse.Site

### PR DESCRIPTION
**Pull request by beckylino**

### CODE REVIEW

File: source/Glimpse.Site/Areas/api/Controllers/VersionController.cs [hunk 1]
```diff
@@ -1,23 +1,33 @@
...
+
 
 namespace Glimpse.Site.Areas.api.Controllers
-{ 
+{
     [HttpHeader("Access-Control-Allow-Origin", "*")]
     public class VersionController : ApiController
-    { 
+    {
+        private TelemetryClient telemetry = new TelemetryClient();
```
Comment (avanderhoorn) [inline source/Glimpse.Site/Areas/api/Controllers/VersionController.cs]: I know this is picky but just from a style perspective can there be a new line between the private variables and the action?

File: source/Glimpse.Site/Areas/api/Controllers/VersionController.cs [hunk 2]
```diff
@@ -1,23 +1,33 @@
 ﻿using System;
 using System.Web.Http;
+using System.Collections.Generic;
 using Glimpse.Package;
 using Glimpse.Site.Framework;
+using Microsoft.ApplicationInsights;
+
```
Comment (avanderhoorn) [inline source/Glimpse.Site/Areas/api/Controllers/VersionController.cs]: Again being picky, if we can remove the extra line here?

File: source/Glimpse.Site/Areas/api/Controllers/VersionController.cs [hunk 3]
```diff
@@ -29,7 +39,7 @@ private string GenerateViewUri(Uri uri, ReleaseQueryInfo result)
                 spacer = "&";
             }
 
-            return  String.Format("{0}://{1}/version/check?{2}", uri.Scheme, uri.Authority, queryString);
+            return String.Format("{0}://{1}/version/check?{2}", uri.Scheme, uri.Authority, queryString);
```
Comment (avanderhoorn) [inline source/Glimpse.Site/Areas/api/Controllers/VersionController.cs]: Thanks!

File: source/Glimpse.Site/Views/Shared/_Layout.cshtml [hunk 4]
```diff
@@ -13,6 +13,18 @@
...
+    var appInsights=window.appInsights||function(config){
+        function s(config){t[config]=function(){var i=arguments;t.queue.push(function(){t[config].apply(t,i)})}}var t={config:config},r=document,f=window,e="script",o=r.createElement(e),i,u;for(o.src=config.url||"//az416426.vo.msecnd.net/scripts/a/ai.0.js",r.getElementsByTagName(e)[0].parentNode.appendChild(o),t.cookie=r.cookie,t.queue=[],i=["Event","Exception","Metric","PageView","Trace"];i.length;)s("track"+i.pop());return config.disableExceptionTracking||(i="onerror",s("_"+i),u=f[i],f[i]=function(config,r,f,e,o){var s=u&&u(config,r,f,e,o);return s!==!0&&t["_"+i](config,r,f,e,o),s}),t
+    }({
+        instrumentationKey:"a052ff85-1c8e-4175-97b1-0b4fb9b09600"
+    });
+
+    window.appInsights=appInsights;
+    appInsights.trackPageView();
+    </script>
+
```
Comment (avanderhoorn) [inline source/Glimpse.Site/Views/Shared/_Layout.cshtml]: Couple of things. Does this need to go in the head? Also does it make sense for this script to be wrapped up in a bundle called `~/bundles/headsitejs` like what we have [here](https://github.com/Glimpse/Glimpse.Site/blob/9c03bd60c08497ace264d35fd3df0e74a8003ada/source/Glimpse.Site/App_Start/BundleConfig.cs#L11-L14)

**Comment by CGijbels** [comment]
I've a general remark regarding the configuration of the Instrumentation key for Application Insights.

The thing with the Instrumentation key being specified in the `ApplicationInsights.config` and directly in the `_Layout.cshtml` is that it is a lot harder to switch those over when moving through different environments like local development/test/.../production.

You might wonder why this is _an issue_? Well the number one reason I have on my list is that I don't like my production telemetry information being mixed with test or even worse local development metrics, especially when exceptions are being tracked. Just imagine building a notification system on top of those metrics and trackings and getting alerts on your mobile phone because somebody is generating a bunch of exceptions while developing a new feature :stuck_out_tongue_winking_eye: 

So I would suggest to remove the Instrumentation key from the applicationinsights.config file and add it programmatically based on a configuration setting. You can then add the following code to the application's startup code

`TelemetryConfiguration.Active.InstrumentationKey = ConfigurationManager.AppSettings["ApplicationInsights:InstrumentationKey"];` and the same for the `_Layout.cshtml` file

That way in development you can keep the existing one (although where is it currently pointing to?) and in production you can change the value in the azure websites configuration, so no need to include it in source control :wink: (although the production instrumentation key is not secret as it will still be exposed by including it in the `_Layout.cshtml` , but that is a different discussion which I already had with the Application Insights team). But at least, without malicious intent, the production metrics will remain useful that way.

Just my 2 :moneybag:

File: source/Glimpse.Site/Areas/api/Controllers/VersionController.cs (hunk 1 above)
Comment (beckylino) [inline source/Glimpse.Site/Areas/api/Controllers/VersionController.cs]: Of course :) Styling is important. Should be fixed now.

**Comment by SergeyKanzhelev** [comment]
@CGijbels thanks for your feedback. Yes, in many cases you want to share instrumentation key. We discussed it with @beckylino and for now we placed snippet in a separate file as @avanderhoorn suggested to download it using bundler. This will speed up loading, but will move instrumentation key even deeper into the javascript file. We probably can initialize instrumentation key right after bundler call. 

I'd suggest to keep and upload it this way for now, see how telemetry looks like and than address feedback on telemetry and investigate how to share the instrumentation key with the server side key properly.

BTW, instead of using app settings - you can do something like [this](http://apmtips.com/blog/2014/11/28/javascript-snippet-and-instrumentation-key/).

**Comment by avanderhoorn** [comment]
@SergeyKanzhelev If you look at https://www.visualstudio.com you will see that the bootstrapper code in in a separate file and the key is in the body of the page.

**Comment by CGijbels** [comment]
@SergeyKanzhelev using the `TelemetryConfiguration.Active.InstrumentationKey` is indeed a lot better than using the app settings, thanks for that tip, now it only needs to be set at one place

Either way, how the instrumentation key is sent down the client is one thing, but I wonder if you agree on the fact that the instrumentation key should not be set inside the `ApplicationInsights.config` file, but programmatically by reading it from the app settings, so that at least 2 different instrumentation keys and therefor different Application Insights App will be used to separate the production telemetry data from the other environments....

File: source/Glimpse.Site/App_Start/BundleConfig.cs [hunk 5]
```diff
@@ -8,6 +8,9 @@ public class BundleConfig
         // For more information on bundling, visit http://go.microsoft.com/fwlink/?LinkId=301862
         public static void RegisterBundles(BundleCollection bundles)
         {
+            bundles.Add(new ScriptBundle("~/bundles/headsitejs").Include(
+                "~/Scripts/ApplicationInsightsSnippet.js"));
```
Comment (avanderhoorn) [inline source/Glimpse.Site/App_Start/BundleConfig.cs]: Can this be inline with the rest of the bundles?

Comment (SergeyKanzhelev) [inline source/Glimpse.Site/App_Start/BundleConfig.cs]: We recommend to put it into the head to catch errors happening during pages and scripts parsing and have a chance to send telemetry if user leaving the page before the full loading. We can move it to the bottom of body tag, but I'd advice to keep it this way.

**Comment by SergeyKanzhelev** [comment]
@avanderhoorn wow, visualstudio is still using "old" Application Insights - one integrated into Visual Studio Portal. It will be [shut down](http://azure.microsoft.com/blog/2015/08/06/transitioning-application-insights-from-the-vso-portal-to-the-azure-portal/) very soon.

However the approach is exactly the same as we can implement here. They actually set global before loading the script and use this variable to initialize Application Insights. We can do it with Glimpse as well. I thought we may have more feedback and telemetry needs once we will look at production data. So wanted to fix it in one shot. We can send an update right away though...

File: source/Glimpse.Site/App_Start/BundleConfig.cs (hunk 5 above)
Comment (avanderhoorn) [inline source/Glimpse.Site/App_Start/BundleConfig.cs]: Thats what I thought but wanted to confirm

**Pull request merged**